Neo4j.Driver/Neo4j.Driver/Internal/Routing/ClusterDiscoveryManager.cs

```diff
@@ -61,46 +61,95 @@ public void Rediscovery()
                 {
                     var result = session.Run(DiscoveryProcedure);
                     var record = result.Single();
+                    ParseDiscoveryResult(record);
+                }
+            }
+            catch (Exception e)
+            {
+                HandleDiscoveryException(e);
+            }
+
+            if (!Readers.Any() || !Routers.Any())
+            {
+                throw new ProtocolException(
+                    $"Invalid discovery result: discovered {Routers.Count()} routers, " +
+                    $"{Writers.Count()} writers and {Readers.Count()} readers.");
+            }
+        }
 
-                    foreach (var servers in record["servers"].As<List<Dictionary<string, object>>>())
-                    {
-                        var addresses = servers["addresses"].As<List<string>>();
-                        var role = servers["role"].As<string>();
-                        switch (role)
-                        {
-                            case "READ":
-                                Readers = addresses.Select(BoltRoutingUri).ToArray();
-                                break;
-                            case "WRITE":
-                                Writers = addresses.Select(BoltRoutingUri).ToArray();
-                                break;
-                            case "ROUTE":
-                                Routers = addresses.Select(BoltRoutingUri).ToArray();
-                                break;
-                        }
-                    }
-                    ExpireAfterSeconds = record["ttl"].As<long>();
+        public async Task RediscoveryAsync()
+        {
+            var provider = new SingleConnectionBasedConnectionProvider(_conn);
+            var session = new Session(provider, _logger);
+            try
+            {
+                var result = await session.RunAsync(DiscoveryProcedure).ConfigureAwait(false);
+                await result.ReadAsync().ConfigureAwait(false);
```

Comment:
It would be better to standardize this usage as 
```csharp
if (await result.ReadAsync().ConfigureAwait(false)) {

} else {
   throw ....
}
```
or negation of it.